src/pip/_internal/download.py

```diff
@@ -939,7 +939,16 @@ def unpack_file_url(
     if is_dir_url(link):
         if os.path.isdir(location):
             rmtree(location)
-        shutil.copytree(link_path, location, symlinks=True)
+        shutil.copytree(link_path,
+                        location,
+                        symlinks=True,
+                        # Pulling in those directories can potentially
+                        # be very slow.
+                        # see dicsussion at:
+                        #   https://github.com/pypa/pip/issues/2195
+                        #   https://github.com/pypa/pip/pull/2196
+                        ignore=shutil.ignore_patterns('.tox', '.nox'))
```

Comment:
Can we limit this to only exclude `.tox` and `.nox` directories that are in the root path? I could see a project that has:

1. `tests` included in their package
2. data directories within tests that have `.tox` and `.nox` subdirectories for specific test cases

getting bit by the current behavior.

Reply:
typically those .nox and .tox directories are transient and are generated automatically when the user run them for the first time.
They are a cache. excluding them just means that the cache will have to be generated again when the test is ran.
Can you think of a specific use case that actually assume those directories already exist?

Reply:
The use case would be:

1. I make a CLI tool `pkg`, and put my tests in `pkg/tests`.
2. I want to behave differently inside a project that has a `.tox` directory (for example, I don't want my tool to package up the `.tox` directory) so I make a test case for it
3. For my test case I create `pkg/tests/data/example_project_with_tox/.tox/abc` and the test itself checks that when I package up `example_project_with_tox` it does not include the `.tox` directory
4. To run my tests I `pip install . && python -m pkg.tests` and my tests pass but maybe only because pip did not copy the `pkg/tests/data/example_project_with_tox/.tox` directory in the first place

Reply:
When I write the test code I will ensure that it the code only operates on .tox and .nox in the root of the project.